pkg/file/redundancy/getter/getter.go

```diff
@@ -256,34 +243,32 @@ func (g *decoder) runStrategy(ctx context.Context, s Strategy) error {
 
 	c := make(chan error, len(m))
 
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
 	for _, i := range m {
-		g.wg.Add(1)
 		go func(i int) {
-			defer g.wg.Done()
 			c <- g.fetch(ctx, i, false)
 		}(i)
 	}
 
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-c:
-			if g.fetchedCnt.Load() >= int32(g.shardCnt) {
-				return nil
-			}
-			if g.failedCnt.Load() > int32(allowedErrs) {
-				return errStrategyFailed
-			}
+	for range c {
+		if g.fetchedCnt.Load() >= int32(g.shardCnt) {
```

Comment:
every time c returns, it's guaranteed that either fetchedCnt or failedCnt is already incremented (inside fetch) so eventually one of the conditions are met. 

if m is less than shard + parity, then a previous run of fetches has already finished and incremented the counters.

Reply:
think about this once more. Because of the single flight pattern and Get calls interleaving prefetch fetches, there is a decent probability that the loop will not terminate. All you need to imagine is that the shardCnt-th successful fetch is.
I think we should trigger closing the initRecovery channel when we  increment the count (in one atomic call)